### PR TITLE
Start working on mashup (#75)

### PR DESCRIPTION
**Pull request by krofna**
This is a draft.

### CODE REVIEW

File: tle/cogs/codeforces.py [hunk 1]
```diff
@@ -72,6 +72,38 @@ def __init__(self, bot):
...
+    @commands.command(brief='Create a mashup')
+    @cf_common.user_guard(group='gitgud')
+    async def mashup(self, ctx, *handles: str):
+        handles = handles or ('!' + str(ctx.author),)
+        handles = await cf_common.resolve_handles(ctx, self.converter, handles)
+        resp = [await cf.user.status(handle=handle) for handle in handles]
+        submissions = [sub for user in resp for sub in user]
+        solved = {sub.problem.name for sub in submissions}
+
+        try:
```
Comment (meooow25) [inline tle/cogs/codeforces.py]: try-except not necessary here (since #69)

File: tle/cogs/codeforces.py [hunk 2]
```diff
@@ -72,6 +72,38 @@ def __init__(self, bot):
...
+            info = await cf.user.info(handles=handles)
+        except cf.CodeforcesApiError:
+            await ctx.send('Codeforces API error.')
+            return
+
+        rating = int(round(sum([user.rating or 1500 for user in info]) / len(handles), -2))
```
Comment (meooow25) [inline tle/cogs/codeforces.py]: [] is redundant here.

File: tle/cogs/codeforces.py [hunk 3]
```diff
@@ -72,6 +72,38 @@ def __init__(self, bot):
...
+        problems = [prob for prob in cf_common.cache2.problem_cache.problems
+                    if prob.rating >= rating - 100 and prob.rating <= rating + 100 and prob.name not in solved
+                    and not any([cf_common.is_contest_writer(prob.contestId, handle) for handle in handles])]
```
Comment (meooow25) [inline tle/cogs/codeforces.py]: [] is redundant here too.

File: tle/cogs/codeforces.py [hunk 4]
```diff
@@ -72,6 +72,38 @@ def __init__(self, bot):
...
+
+        try:
+            info = await cf.user.info(handles=handles)
+        except cf.CodeforcesApiError:
+            await ctx.send('Codeforces API error.')
+            return
+
+        rating = int(round(sum([user.rating or 1500 for user in info]) / len(handles), -2))
+        problems = [prob for prob in cf_common.cache2.problem_cache.problems
+                    if prob.rating >= rating - 100 and prob.rating <= rating + 100 and prob.name not in solved
```
Comment (meooow25) [inline tle/cogs/codeforces.py]: `abs(rating - problem.rating) <= 100`

File: tle/cogs/codeforces.py [hunk 5]
```diff
@@ -72,6 +72,38 @@ def __init__(self, bot):
             embed.add_field(name='Matched tags', value=tagslist)
         await ctx.send(f'Recommended problem for `{handle}`', embed=embed)
 
+    @commands.command(brief='Create a mashup')
+    @cf_common.user_guard(group='gitgud')
```
Comment (meooow25) [inline tle/cogs/codeforces.py]: I don't think user_guard is needed here.

File: tle/cogs/codeforces.py [hunk 6]
```diff
@@ -72,6 +72,38 @@ def __init__(self, bot):
             embed.add_field(name='Matched tags', value=tagslist)
         await ctx.send(f'Recommended problem for `{handle}`', embed=embed)
 
+    @commands.command(brief='Create a mashup')
```
Comment (meooow25) [inline tle/cogs/codeforces.py]: Probably you intend to add it later, but don't forget the help string.

File: tle/cogs/codeforces.py [hunk 7]
```diff
@@ -72,6 +72,38 @@ def __init__(self, bot):
...
+        if len(problems) < 4:
+            await ctx.send('Problems not found within the search parameters')
+            return
+
+        problems.sort(key=lambda problem: -cf_common.cache2.contest_cache.get_contest(
+            problem.contestId).startTimeSeconds)
+
+        msg = ''
+        for i in range(4):
+            msg += f'{"ABCD"[i]}: {problems[i].name} - <{problems[i].url}>\n'
```
Comment (meooow25) [inline tle/cogs/codeforces.py]: I know it's just 4 iterations but maybe don't use `+=` on strings.
You could even one-line it: `msg = '\n'.join(whatever for prob in problems[4])`

File: tle/cogs/codeforces.py [hunk 8]
```diff
@@ -72,6 +72,38 @@ def __init__(self, bot):
...
+        await ctx.send(msg)
```
Comment (meooow25) [inline tle/cogs/codeforces.py]: Use an embed instead for maximum prettiness.

Comment (krofna) [reply tle/cogs/codeforces.py]: waturr

Comment (krofna) [reply tle/cogs/codeforces.py]: meh just merge this and add an embed later @meooow25

**Pull request merged**